pkg/cvo/upgradeable.go

```diff
@@ -234,6 +247,126 @@ func (check *clusterManifestDeleteInProgressUpgradeable) Check() *configv1.Clust
 	return nil
 }
 
+func gateApplicableToCurrentVersion(gateName string, currentVersion string) (bool, error) {
+	var applicable bool
+	if ackVersion := adminAckGateRegexp.FindString(gateName); ackVersion == "" {
+		return false, fmt.Errorf("%s configmap gate name %s has invalid format; must begin with 'ack-X.Y'",
+			internal.AdminGatesConfigMap, gateName)
+	} else {
+		parts := strings.Split(ackVersion, "-")
+		if strings.HasPrefix(currentVersion, parts[1]) {
+			applicable = true
+		}
+	}
+	return applicable, nil
+}
+
+func checkAdminGate(gate string, currentVersion string,
+	gateConfigmap *corev1.ConfigMap, ackConfigmap *corev1.ConfigMap,
+	ackConfigmapErr error) *configv1.ClusterOperatorStatusCondition {
+
+	if applies, err := gateApplicableToCurrentVersion(gate, currentVersion); err == nil {
+		if !applies {
+			return nil
+		}
+	} else {
+		klog.Error(err)
+		return &configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.ClusterStatusConditionType("UpgradeableAdminAckRequired"),
+			Status:  configv1.ConditionFalse,
+			Reason:  "AdminAckConfigMapIssue",
+			Message: err.Error(),
+		}
+	}
+	var errorMessage string
+	var gateMessage string
+	if val, ok := gateConfigmap.Data[gate]; !ok {
+		errorMessage = fmt.Sprintf("%s configmap does not contain gate %s.", internal.AdminGatesConfigMap, gate)
+	} else if val == "" {
+		errorMessage = fmt.Sprintf("%s configmap gate %s has no value.", internal.AdminGatesConfigMap, gate)
+	} else {
+		gateMessage = val
+	}
+	if ackConfigmapErr != nil {
+		if apierrors.IsNotFound(ackConfigmapErr) {
+			errorMessage = strings.Trim(errorMessage+" "+fmt.Sprintf("%s configmap not found.", internal.AdminAcksConfigMap), " ")
+		} else if ackConfigmapErr != nil {
+			errorMessage = strings.Trim(errorMessage+" "+
+				fmt.Sprintf("Unable to access configmap %s, err=%v.", internal.AdminAcksConfigMap, ackConfigmapErr), " ")
```

Comment:
This feels like it should be handled in the caller, or the ConfigMap fetch should be pushed down into this function.  Is there a reason to have the fetcher in one location, and the failed-fetch handlers in another?

Reply:
I did it this way so gates can be added easily:

`conds []configv1.ClusterOperatorStatusCondition`
`conds[0] = checkAdminGate("ack-4.8-kube-122-api-removals-in-4.9", currentVersion, gateCm, ackCm,err)`
`... = checkAdminGate("ack-4.9-foo", currentVersion, gateCm, ackCm, err)`

Reply:
Can't we:

```go
ackCm, err := check.adminAcksLister.Get(internal.AdminAcksConfigMap)
// handle err, and fast-fail with these messages as needed.
ackErrors := []string{}
for _, ack := range []string{"well", "known", "acks", ...} {
  if err := checkAdminGate(ack, currentVersion, gateCm, ackCm); err != nil {
    ackErrors = append(ackErrors, err)
  }
}
return someAggregatingThing(ackErrors)
```

Reply:
Sure. I was just trying to show why I split the logic up as I did not that I would necessarily implement it exactly like that.